src/AspNet.Security.OAuth.WeChat/WeChatAuthenticationHandler.cs

```diff
@@ -0,0 +1,147 @@
+                            /* Status: */ response.StatusCode,
+                            /* Headers: */ response.Headers.ToString(),
+                            /* Body: */ await response.Content.ReadAsStringAsync());
+
+                        throw new HttpRequestException("An error occurred while retrieving user information.");
+                    }
+
+                    payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+                    if (!string.IsNullOrEmpty(payload.Value<string>("errcode")))
+                    {
+                        Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                        "returned a {Status} response with the following payload: {Headers} {Body}.",
+                            /* Status: */ response.StatusCode,
+                            /* Headers: */ response.Headers.ToString(),
+                            /* Body: */ await response.Content.ReadAsStringAsync());
+
+                        throw new HttpRequestException("An error occurred while retrieving user information.");
+                    }
+
+                    break;
+                default:
+                    throw new InvalidOperationException("Invalid scope");
+            }
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options,
+                Backchannel, tokens, payload);
+            context.RunClaimActions(payload);
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(string code, string redirectUri)
+        {
+            var address = QueryHelpers.AddQueryString(Options.TokenEndpoint, new Dictionary<string, string>
+            {
+                ["appid"] = Options.ClientId,
+                ["secret"] = Options.ClientSecret,
+                ["code"] = code,
+                ["grant_type"] = "authorization_code"
+            });
+
+            var response = await Backchannel.GetAsync(address);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving an access token: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
```

Comment:
`Exception` shouldn't really be used to throw exceptions as it's so non-specific. Is there a better way you can achieve this?